File: src/test/java/com/djdapz/stubhub/controller/ListingControllerTest.java
package com.djdapz.stubhub.controller;

import com.djdapz.stubhub.domain.ProcessedListing;
import com.djdapz.stubhub.service.ListingService;
import lombok.experimental.FieldDefaults;
import org.junit.Before;
import org.junit.Test;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.util.Arrays;
import java.util.List;

import static com.djdapz.stubhub.config.JsonConfig.asObject;
import static com.djdapz.stubhub.util.Random.randomInt;
import static com.djdapz.stubhub.util.Random.randomProcessedListing;
import static lombok.AccessLevel.PRIVATE;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@FieldDefaults(level = PRIVATE, makeFinal = true)
public class ListingControllerTest {

    ListingService listingService = mock(ListingService.class);
    ListingController subject = new ListingController(listingService);

    MockMvc mockMvc = MockMvcBuilders.standaloneSetup(subject).build();
    Integer eventId = randomInt();

    List<ProcessedListing> expectedListings = Arrays.asList(randomProcessedListing(), randomProcessedListing());

    @Before
    public void setUp() {
        when(listingService.getListingFor(any())).thenReturn(expectedListings);
    }

    @Test
    public void shouldReturn200() throws Exception {
        mockMvc.perform(
                get("/listings")
                        .param("eventId", eventId.toString())
        ).andExpect(status().isOk());
    }


    @Test
    public void shouldDelegateToListingSservice() throws Exception {
        mockMvc.perform(
                get("/listings")
                        .param("eventId", eventId.toString())
        );

        verify(listingService).getListingFor(eventId);
    }

    @Test
    public void shoulReturnListOfListinfs() throws Exception {
        String eventId = mockMvc.perform(
                get("/listings")
                        .param("eventId", this.eventId.toString()))
                .andReturn()
                .getResponse()
                .getContentAsString();

        ProcessedListing[] actual = asObject(eventId, ProcessedListing[].class);

        assertThat(actual).containsExactlyElementsOf(expectedListings);
    }
}

File: src/test/java/com/djdapz/stubhub/controller/ListingControllerTest.kt
package com.djdapz.stubhub.controller

import com.djdapz.stubhub.config.JsonConfig.Companion.asObject
import com.djdapz.stubhub.domain.ProcessedListing
import com.djdapz.stubhub.service.ListingService
import com.djdapz.stubhub.util.randomInt
import com.djdapz.stubhub.util.randomProcessedListing
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.doReturn
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import java.util.Arrays.asList

class ListingControllerTest {
    private val eventId: Int = randomInt()
    private val expectedListings = asList(randomProcessedListing(), randomProcessedListing())

    private val listingService = mock<ListingService> {
        on { getListingFor(any()) } doReturn expectedListings
    }
    private val subject = ListingController(listingService)
    private val mockMvc = MockMvcBuilders.standaloneSetup(subject).build()

    @Test
    @Throws(Exception::class)
    fun shouldDelegateToListingService() {
        mockMvc.perform(
                get("/listings")
                        .param("eventId", eventId.toString())
        )

        verify(listingService).getListingFor(eventId)
    }

    @Test
    @Throws(Exception::class)
    fun shouldReturnListOfListings() {
        val listingsJson = mockMvc.perform(
                get("/listings")
                        .param("eventId", this.eventId.toString()))
                .andReturn()
                .response
                .contentAsString

        val actual = asObject(listingsJson, Array<ProcessedListing>::class.java)

        assertThat(actual).containsExactlyElementsOf(expectedListings)
    }
}